cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,309 @@
+Equality is also used implicitly by literal maps in node and relationship patterns, since such literal maps are merely a shorthand notation for equality predicates.
+[[orderability-con,orderability (concept)]]*Orderability*:: <<orderability-def,Orderability>> revolves around the `ORDER BY` clause, and defines the semantics it needs to be able to properly sort values passed to it.
+[[equivalence-con,equivalence (concept)]]*Equivalence*:: <<equivalence-def,Equivalence>> revolves around the `DISTINCT` modifier and grouping in projection clauses (`WITH`, `RETURN`), and defines the semantics they need to properly determine if two values are the same.
+
+==== The meaning of `null`
+
+For the following discussion, it is helpful to clarify the meaning of `null`. In Cypher, a `null` value has one of two meanings, depending on the context in which it occurs:
+
+[horizontal]
+[[unknown-null,unknown `null`]]*Unknown*:: An "unknown" `null` is taken to be a placefolder for an arbitrary but unknown value. When evaluating predicates, an "unknown" `null` is the `maybe` truth value of ternary logic. For node and relationship properties, an "unknown" `null` is a value that is definite in the real world but has not been stored in the graph. Since in these cases, two "unknown" `null` values stand for arbitrary but definite values in the real world, two "unknown" `null` values should never be treated as certainly being the same value.
+[[missing-null,missing `null`]]*Missing*:: A "missing" `null` is taken to be a marker for the absence of a value. In the context of updating node properties from a map, a "missing" `null` is used to mark properties that are to be removed. In the context of `DISTINCT` and grouping, a "missing" `null` value is used as grouping key for all records that miss a more specific value. Since in these cases, two "missing" `null` values represent the same concept, they should always be treated as the same value.
+
+[[regular-maps]]
+==== Regular maps
+
+Cypher today has one supertype `MAP` for all map values. This includes nodes (of subtype `NODE`), relationships (of subtype `RELATIONSHIP`), and any other map (not captured by a subtype of `MAP`). For the purpose of this document, we define a regular map to be any value of type `MAP` that is neither a `NODE` nor a `RELATIONSHIP`.
+
+<<comparability-equality,comparability and equality>>
+=== Comparability and equality
+
+We propose that comparability and equality should be aligned with each other, i.e.
+
+`expr1 = expr2` if and only if `expr1 >= expr2 && expr1 \<= expr2`.
+
+Comparability and equality use <<unknown-null,"unknown" `null` values>>.
+
+[[incomparable,incomparable]]
+==== Incomparability
+
+We define that a value is incomparable if and only if any comparison or equality test involving this value evaluates to `null`.
```

Comment:
What is the purpose of this definition? It seems to define every single value as incomparable, seeing as comparing any value to `null` will evaluate to `null`. If it is motivated, please provide a motivation.